File: IcarianCS/src/Rendering/UI/UIElement.cs
// Icarian Engine - C# Game Engine
// 
// License at end of file.

using IcarianEngine.Maths;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

#include "EngineUIElementInterop.h"
#include "EngineUIElementInteropStuctures.h"
#include "InteropBinding.h"

ENGINE_UIELEMENT_EXPORT_TABLE(IOP_BIND_FUNCTION);

namespace IcarianEngine.Rendering.UI
{
    public class UIElement : IDestroy
    {
        static readonly ConcurrentDictionary<uint, UIElement> s_elementLookup = new ConcurrentDictionary<uint, UIElement>();

        /// <summary>
        /// Delegate for UI events
        /// </summary>
        /// <param name="a_canvas">The <see cref="IcarianEngine.Rendering.UI.Canvas" /> the event is for</param>
        /// <param name="a_element">The <see cref="IcarianEngine.Rendering.UI.UIElement" /> the event is for</param>
        public delegate void UIEvent(Canvas a_canvas, UIElement a_element);

        uint   m_bufferAddr = uint.MaxValue;
        string m_name;

        /// <summary>
        /// Delegate for normal events
        /// </summary>
        public UIEvent OnNormal = null;
        /// <summary>
        /// Delegate for hover events
        /// </summary>
        public UIEvent OnHover = null;
        /// <summary>
        /// Delegate for press events
        /// </summary>
        public UIEvent OnPressed = null;
        /// <summary>
        /// Delegate for release events
        /// </summary>
        public UIEvent OnReleased = null;

        /// <summary>
        /// Whether or not the UIElement is Disposed/Finalised
        /// </summary>
        public bool IsDisposed
        {
            get
            {
                return m_bufferAddr == uint.MaxValue;
            }
        }

        internal uint BufferAddr
        {
            get
            {
                return m_bufferAddr;
            }
            set
            {
                m_bufferAddr = value;
            }
        }

        /// <summary>
        /// The name of the UIElement
        /// </summary>
        public string Name
        {
            get
            {
                return m_name;
            }
            internal set
            {
                m_name = value;
            }
        }

        /// <summary>
        /// The current state of the UIElement
        /// </summary>
        public ElementState ElementState
        {
            get
            {
                return (ElementState)UIElementInterop.GetElementState(m_bufferAddr);
            }
        }

        /// <summary>
        /// The position of the UIElement in <see cref="IcarianEngine.Rendering.UI.Canvas" /> space
        /// </summary>
        public Vector2 Position
        {
            get
            {
                // Get stack corruption with return value refer to EngineUIElementInterop.h
                IntPtr ptr = Marshal.AllocHGlobal(Marshal.SizeOf<Vector2>());

                UIElementInterop.GetPosition(m_bufferAddr, ptr);
                Vector2 val = Marshal.PtrToStructure<Vector2>(ptr);
                Marshal.FreeHGlobal(ptr);

                return val;
            }
            set
            {
                UIElementInterop.SetPosition(m_bufferAddr, value);
            }
        }
        /// <summary>
        /// The size of the UIElement in <see cref="IcarianEngine.Rendering.UI.Canvas" /> space
        /// </summary>
        public Vector2 Size
        {
            get
            {
                // Get stack corruption with return value refer to EngineUIElementInterop.h
                IntPtr ptr = Marshal.AllocHGlobal(Marshal.SizeOf<Vector2>());

                UIElementInterop.GetSize(m_bufferAddr, ptr);
                Vector2 val = Marshal.PtrToStructure<Vector2>(ptr);
                Marshal.FreeHGlobal(ptr);

                return val;
            }
            set
            {
                UIElementInterop.SetSize(m_bufferAddr, value);
            }
        }

        /// <summary>
        /// The colour of the UIElement
        /// </summary>
        public Color Color
        {
            get
            {
                return UIElementInterop.GetColor(m_bufferAddr).ToColor();
            }
            set
            {
                UIElementInterop.SetColor(m_bufferAddr, value.ToVector4());
            }
        }

        /// <summary>
        /// The X anchor point for the UIElement
        /// </summary>
        public UIXAnchor XAnchor
        {
            get
            {
                return (UIXAnchor)UIElementInterop.GetElementXAnchor(m_bufferAddr);
            }
            set
            {
                UIElementInterop.SetElementXAnchor(m_bufferAddr, (uint)value);
            }
        }
        /// <summary>
        /// The Y anchor point for the UIElement
        /// </summary>
        public UIYAnchor YAnchor
        {
            get
            {
                return (UIYAnchor)UIElementInterop.GetElementYAnchor(m_bufferAddr);
            }
            set
            {
                UIElementInterop.SetElementYAnchor(m_bufferAddr, (uint)value);
            }
        }
        
        /// <summary>
        /// The children of the UIElement
        /// </summary>
        public IEnumerable<UIElement> Children
        {
            get
            {
                uint[] children = UIElementInterop.GetChildren(m_bufferAddr);

                foreach (uint child in children)
                {
                    if (s_elementLookup.ContainsKey(child))
                    {
                        yield return s_elementLookup[child];
                    }
                }
            }
        }

        public UIElement()
        {
            m_bufferAddr = UIElementInterop.CreateUIElement();

            AddLookup(m_bufferAddr, this);
        }
        protected internal UIElement(uint a_bufferAddr)
        {
            m_bufferAddr = a_bufferAddr;

            AddLookup(m_bufferAddr, this);
        }

        /// @cond INTERNAL

        internal void AddLookup(uint a_addr, UIElement a_element)
        {
            s_elementLookup.TryAdd(a_addr, a_element);
        }
        internal void RemoveLookup(uint a_addr)
        {
            s_elementLookup.TryRemove(a_addr, out UIElement _);
        }

        internal static UIElement GetUIElement(uint a_addr)
        {
            if (s_elementLookup.ContainsKey(a_addr))
            {
                return s_elementLookup[a_addr];
            }

            return null;
        }

        /// @endcond

        /// <summary>
        /// Gets the child of the UIElement with name recursively
        /// </summary>
        /// <param name="a_name">The name of the child to get</param>
        /// <returns>The child UIElement. Null on failure</returns>
        public UIElement GetNamedChild(string a_name)
        {
            foreach (UIElement child in Children)
            {
                if (child.Name == a_name)
                {
                    return child;
                }

                UIElement element = child.GetNamedChild(a_name);
                if (element != null)
                {
                    return element;
                }
            }

            return null;
        }
        /// <summary>
        /// Gets the child of the UIElement of type T with name recursively
        /// </summary>
        /// <param name="a_name">The name of the child to get</param>
        /// <returns>The child UIElement. Null on failure</returns>
        public T GetNamedChild<T>(string a_name) where T : UIElement
        {
            return GetNamedChild(a_name) as T;
        }

        /// <summary>
        /// Adds a child to the UIElement
        /// </summary>
        /// <param name="a_child">The child to add</param>
        public void AddChild(UIElement a_child)
        {
            if (a_child == null)
            {
                Logger.IcarianWarning("Null UIElement");

                return;
            }

            UIElementInterop.AddChildElement(BufferAddr, a_child.BufferAddr);
        }
        /// <summary>
        /// Removes a child from the UIElement
        /// </summary>
        /// <param name="a_child">The child to remove</param>
        public void RemoveChild(UIElement a_child)
        {
            if (a_child == null)
            {
                Logger.IcarianWarning("Null UIElement");

                return;
            }

            UIElementInterop.RemoveChildElement(BufferAddr, a_child.BufferAddr);
        }

        static void OnNormalS(uint a_canvasAddr, uint a_elementAddr)
        {
            Canvas canvas = Canvas.GetCanvas(a_canvasAddr);
            UIElement element = GetUIElement(a_elementAddr);

            if (element.OnNormal != null)
            {
                element.OnNormal.Invoke(canvas, element);
            }
        }
        static void OnHoverS(uint a_canvasAddr, uint a_elementAddr)
        {
            Canvas canvas = Canvas.GetCanvas(a_canvasAddr);
            UIElement element = GetUIElement(a_elementAddr);

            if (element.OnHover != null)
            {
                element.OnHover.Invoke(canvas, element);
            }
        }
        static void OnPressedS(uint a_canvasAddr, uint a_elementAddr)
        {
            Canvas canvas = Canvas.GetCanvas(a_canvasAddr);
            UIElement element = GetUIElement(a_elementAddr);

            if (element.OnPressed != null)
            {
                element.OnPressed.Invoke(canvas, element);
            }
        }
        static void OnReleasedS(uint a_canvasAddr, uint a_elementAddr)
        {
            Canvas canvas = Canvas.GetCanvas(a_canvasAddr);
            UIElement element = GetUIElement(a_elementAddr);

            if (element.OnReleased != null)
            {
                element.OnReleased.Invoke(canvas, element);
            }
        }

        /// <summary>
        /// Disposes of the UIElement
        /// </summary>
        public void Dispose()
        {
            Dispose(true);

            GC.SuppressFinalize(this);
        }
        /// <summary>
        /// Called when the UIElement is Disposed/Finalised
        /// </summary>
        /// <param name="a_disposing">Whether or not it is called from Dispose</param>
        protected virtual void Dispose(bool a_disposing)
        {
            if (m_bufferAddr != uint.MaxValue)
            {
                if (a_disposing)
                {
                    UIElementInterop.DestroyUIElement(m_bufferAddr);

                    RemoveLookup(m_bufferAddr);

                    foreach (UIElement child in Children)
                    {
                        child.Dispose();
                    }
                }
                else
                {
                    Logger.IcarianWarning("UIElement failed to Dispose");
                }

                m_bufferAddr = uint.MaxValue;
            }
            else
            {
                Logger.IcarianError("Multiple Dispose calls on UIElement");
            }
        }
        ~UIElement()
        {
            Dispose(false);
        }
    }
}

// MIT License
// 
// Copyright (c) 2024 River Govers
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

File: IcarianNative/src/UIElement.cpp
// Icarian Engine - C# Game Engine
// 
// License at end of file.

#include "Rendering/UI/UIElement.h"
#include "IcarianError.h"

#include <glm/gtx/matrix_transform_2d.hpp>

UIElement::UIElement()
{
    m_parent = -1;
    m_childCount = 0;
    m_children = nullptr;

    m_xAnchor = UIXAnchor_Middle;
    m_yAnchor = UIYAnchor_Middle;

    m_pos = glm::vec2(0.0f);
    m_size = glm::vec2(100.0f);

    m_color = glm::vec4(1.0f);

    m_state = ElementState_Normal;
}
UIElement::~UIElement()
{
    if (m_children != nullptr)
    {
        delete[] m_children;
        m_children = nullptr;
    }
}

void UIElement::AddChild(uint32_t a_childAddr)
{
    const uint32_t* oldBuffer = m_children;
    IDEFER(
    if (oldBuffer != nullptr)
    {
        delete[] oldBuffer;
    });

    m_children = new uint32_t[m_childCount + 1];
    for (uint32_t i = 0; i < m_childCount; ++i)
    {
        m_children[i] = oldBuffer[i];
    } 

    m_children[m_childCount++] = a_childAddr;
}
void UIElement::RemoveChild(uint32_t a_childAddr)
{
    for (uint32_t i = 0; i < m_childCount; ++i) 
    {
        if (m_children[i] == a_childAddr) 
        {
            --m_childCount;

            for (uint32_t j = i; j < m_childCount; ++j) 
            {
                m_children[j] = m_children[j + 1];
            }

            return;
        }
    }
}

float UIElement::GetXPosition(const CanvasBuffer& a_canvas, const glm::vec2& a_screenSize) const
{
    const float scaledPos = m_pos.x / a_canvas.ReferenceResolution.x;

    switch (m_xAnchor) 
    {
    case UIXAnchor_Left:
    {
        return scaledPos;
    }
    case UIXAnchor_Middle:
    case UIXAnchor_Stretch:
    {
        const float halfSize = GetXSize(a_canvas, a_screenSize) * 0.5f;
        const float finalPos = scaledPos - halfSize;

        if (m_parent != -1)
        {
            const UIElement* parent = UIControl::GetUIElement(m_parent);

            const float parentHalf = parent->GetXSize(a_canvas, a_screenSize) * 0.5f;
            return parentHalf + finalPos;
        }

        return 0.5f + finalPos;
    }
    case UIXAnchor_Right:
    {
        const float size = GetXSize(a_canvas, a_screenSize);
        const float finalPos = scaledPos - size;

        if (m_parent != -1)
        {
            const UIElement* parent = UIControl::GetUIElement(m_parent);

            const float parentSize = parent->GetXSize(a_canvas, a_screenSize);
            return parentSize + finalPos;
        }

        return 1.0f + finalPos;
    }
    }

    IERROR("Invalid UIXAnchor");

    return 0.0f;
}
float UIElement::GetYPosition(const CanvasBuffer& a_canvas, const glm::vec2& a_screenSize) const
{
    const float scaledPos = m_pos.y / a_canvas.ReferenceResolution.y;

    switch (m_yAnchor)
    {
    case UIYAnchor_Top:
    {
        return scaledPos;
    }
    case UIYAnchor_Middle:
    case UIYAnchor_Stretch:
    {
        const float halfSize = GetYSize(a_canvas, a_screenSize);
        const float finalPos = scaledPos - halfSize;

        if (m_parent != -1)
        {
            const UIElement* parent = UIControl::GetUIElement(m_parent);

            const float parentHalf = parent->GetYSize(a_canvas, a_screenSize);
            return parentHalf + finalPos;
        }

        return 0.5f + finalPos;
    }
    case UIYAnchor_Bottom:
    {
        const float size = GetYSize(a_canvas, a_screenSize);
        const float finalPos = scaledPos - size;

        if (m_parent != -1)
        {
            const UIElement* parent = UIControl::GetUIElement(m_parent);

            const float parentSize = parent->GetYSize(a_canvas, a_screenSize);
            return parentSize + finalPos;
        }

        return 1.0f + finalPos;
    }
    }

    IVERIFY("Invalid UIYAnchor");

    return 0.0f;
}

float UIElement::GetXSize(const CanvasBuffer& a_canvas, const glm::vec2& a_screenSize) const
{
    const float scaled = m_size.x / a_canvas.ReferenceResolution.x;

    switch (m_xAnchor) 
    {
    case UIXAnchor_Left:
    case UIXAnchor_Middle:
    case UIXAnchor_Right:
    {
        const float xScale = a_screenSize.x / a_canvas.ReferenceResolution.x;
        return scaled / xScale;
    }
    case UIXAnchor_Stretch:
    {
        return scaled;
    }
    }

    IERROR("Invalid UIXAnchor");

    return 0.0f;
}
float UIElement::GetYSize(const CanvasBuffer& a_canvas, const glm::vec2& a_screenSize) const
{
    const float scaled = m_size.y / a_canvas.ReferenceResolution.y;

    switch (m_yAnchor) 
    {
    case UIYAnchor_Top:
    case UIYAnchor_Middle:
    case UIYAnchor_Bottom:
    {
        const float yScale = a_screenSize.y / a_canvas.ReferenceResolution.y;
        return scaled / yScale;
    }
    case UIYAnchor_Stretch:
    {
        return scaled;
    }
    }

    IERROR("Invalid UIYAnchor");

    return 0.0f;
}

glm::vec2 UIElement::GetCanvasPosition(const CanvasBuffer& a_canvas, const glm::vec2& a_screenSize) const
{
    const glm::vec2 pos = glm::vec2
    (
        GetXPosition(a_canvas, a_screenSize),
        GetYPosition(a_canvas, a_screenSize)
    );

    if (m_parent != -1)
    {
        const UIElement* parent = UIControl::GetUIElement(m_parent);
        const glm::vec2 parentPos = parent->GetCanvasPosition(a_canvas, a_screenSize);

        return parentPos + pos;
    }

    return pos;
}
glm::vec2 UIElement::GetCanvasScale(const CanvasBuffer& a_canvas, const glm::vec2& a_screenSize) const
{
    return glm::vec2
    (
        GetXSize(a_canvas, a_screenSize),
        GetYSize(a_canvas, a_screenSize)
    );
}

void UIElement::Update(RenderEngine* a_renderEngine)
{

}

// MIT License
// 
// Copyright (c) 2024 River Govers
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.